### ISSUE TITLE
Each member of the union type has signatures, but none of those signatures are compatible with each other

### ISSUE THREAD

**Issue by ackvf** [microsoft/TypeScript#33591]
**TypeScript Version:**  3.6.3

**Search Terms:**
array.map, expression is not callable, union type signatures,

**Code**

_The following code is the exact same code as in this codesandbox_
https://codesandbox.io/s/lingering-sun-7itgj

```ts
interface Product_result {
  __typename: 'Product'
  id: number
  description: string | null
}
interface Campaign_result {
  __typename: 'Campaign'
  id: number
  products: (Product_result | null)[] | null
}


interface Product {
  id: number
  description: string
  specific: boolean
}
interface Campaign {
  products: Product[]
}

type CompoundType = Campaign_result | Campaign

/* --- */

const props: { campaign?: Campaign_result } = {}
const emptyCampaign: Campaign = {
  products: []
}

const initialData: CompoundType = props.campaign || emptyCampaign


/*
Cannot invoke an expression whose type lacks a call signature. Type '
    (<U>(callbackfn: (value: Product_result, index: number, array: Product_result[]) => U, thisArg?: any) => U[])
  | (<U>(callbackfn: (value: Product, index: number, array: Product[]) => U, thisArg?: any) => U[])
' has no compatible call signatures.

product inside map is any, but I know `id` is there
*/
initialData.products.map(product => product.id)


// product is { description: string, id: number } - as expected
const product = initialData.products[0] 

// product inside map is { description: string, id: number } - as expected
;(initialData.products as Array<CompoundType['products'][0]>).map(product => product.id)


/* interestingly */

type T01 = Array<CompoundType['products'][0]>   // T01: (Product_result | Product)[]  -  I can actually have mix of both. Products that were fetched and Products added via a Form
type T02 = CompoundType['products']             // T02: Product_result[] | Product[]

declare const v1: T01
v1.map(product => product) // OK

declare const v2: T02
v2.map(product => product) // NOK
```

**Expected behavior:**

I know `id` is there, so this should work
`initialData.products.map(product => product.id)`

This workaround works
`(initialData.products as Array<CompoundType['products'][0]>).map(product => product.id)`

**Actual behavior:**

The map gives a non compatible call signature error (as in the snippet above), and the product inside is any.

**Related**
TypeScript 3.3 - Improved behavior for calling union types
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-3.html#caveats

**Comment by jack-williams** [comment]
I think this is being tracked here #7294.

**Comment by RyanCavanaugh** [comment]
We don't have any way to check this in a way that's well-bounded in terms of analysis time.

Consider if you had written something like this:
```ts
const arr1: string[] | number[] | boolean[] = [];
const arr2: string[] | number[] | boolean[] = [];
const arr3: string[] | number[] | boolean[] = [];

declare function fn(a: string, b: number, c: boolean): void;
declare function fn(a: number, b: number, c: string): void;
declare function fn(a: string, b: boolean, c: boolean): void;
declare function fn(a: number, b: number, c: string): void;

arr1.forEach(a => {
    arr2.forEach(b => {
        arr3.forEach(c => {
            // arr: ????
            const arr = [a, b, c] as const;
            fn(arr[0], arr[1], arr[2]);
        });
    });
});
```
The only way to correctly check this program is to re-check the body of the inner body 3 * 3 * 3 times (!), and TS would need to clear its cache of expression types on each invocation (something which is currently architecturally impossible). It's not even clear what we'd show for `arr`'s type here!

A typical suggestion is some sort of merging of argument types into unions, but this only works in some cases (it's OK for `map` but doesn't make sense for `filter`, for example).

**Comment by ackvf** [comment]
I see, that's understandable. So, is there an official best-practice™ workaround for this or should I resort to manually extract the types like this?
```ts
type T01 = Array<CompoundType['products'][0]>
```

**Comment by whizsid** [comment]
I have solved in that way.

```
type CompoundType = (Campaign_result | Campaign)&{products:  (Product|Product_result)[]}
```

**Comment by Ranguna** [comment]
> ```ts
> const arr1: string[] | number[] | boolean[] = [];
> const arr2: string[] | number[] | boolean[] = [];
> const arr3: string[] | number[] | boolean[] = [];
> 
> declare function fn(a: string, b: number, c: boolean): void;
> declare function fn(a: number, b: number, c: string): void;
> declare function fn(a: string, b: boolean, c: boolean): void;
> declare function fn(a: number, b: number, c: string): void;
> 
> arr1.forEach(a => {
>     arr2.forEach(b => {
>         arr3.forEach(c => {
>             // arr: ????
>             const arr = [a, b, c] as const;
>             fn(arr[0], arr[1], arr[2]);
>         });
>     });
> });
> ```
> It's not even clear what we'd show for arr's type here!

I'm probably missing something here, but to me `arr` would be `[string | number | boolean, string | number | boolean, string | number | boolean]`. This would fail to compile because the first parameter of `fn` doesn't accept a `boolean` value.

**Comment by Fleuv** [comment]
I also experienced this issue with the following use case.

**TypeScript Version:**  3.9.2

**Code**
```ts
interface Base {
    id: number;
}

interface Book extends Base {
    title: string;
    author: Author;
}

interface Author extends Base {
    name: string;
}

interface Models {
    books: Book[];
    authors: Author[];
}

const author1: Author = {
    id: 1,
    name: 'Foo Bar',
};

const data : Models = {
    books: [
        {
            id: 1,
            title: "hello world",
            author: author1,
        }
    ],
    authors: [
        author1,
    ],
};

// This does work
function getBookById(id: number): Book | undefined {
    return data.books.filter((item: Base) => item.id = id)[0];
}

// This is more generic but does not work
/**
 * Error message:
 * 
 * This expression is not callable. Each member of the union type '
 * { 
 *      <S extends Book>
 *          (callbackfn: (value: Book, index: number, array: Book[]) => value is S, thisArg?: any): S[];
 *          (callbackfn: (value: Book, index: number, array: Book[]) => unknown, thisArg?: any): Book[];
 * } | { ...; }
 * ' has signatures, but none of those signatures are compatible with each other.
 */
function getById<T>(type: keyof Models, id: number): T | undefined {
    return data[type].filter((item: Base) => item.id = id)[0];
}
```



<details><summary><b>Output</b></summary>

```ts
"use strict";
const author1 = {
    id: 1,
    name: 'Foo Bar',
};
const data = {
    books: [
        {
            id: 1,
            title: "hello world",
            author: author1,
        }
    ],
    authors: [
        author1,
    ],
};
// This does work
function getBookById(id) {
    return data.books.filter((item) => item.id = id)[0];
}
// This is more generic but does not work
/**
 * Error message:
 *
 * This expression is not callable. Each member of the union type '
 * {
 *      <S extends Book>
 *          (callbackfn: (value: Book, index: number, array: Book[]) => value is S, thisArg?: any): S[];
 *          (callbackfn: (value: Book, index: number, array: Book[]) => unknown, thisArg?: any): Book[];
 * } | { ...; }
 * ' has signatures, but none of those signatures are compatible with each other.
 */
function getById(type, id) {
    return data[type].filter((item) => item.id = id)[0];
}

```


</details>


<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```


</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgEJwM4oN4ChkHLAAmAXMiAK4C2ARtANy4C+uuoksiKqA9rwGtkEAB6QQxDGkw58hMMDAAbCOQxgooAOZNCyOJTAALXlHIBBQyahNW7cNHhJkl46eFiIEqeizI8eiBw1KrI6pogOixsHI7cyACyvMQQSlIBhLT8AhjkfIIA2gC6uoQGblC5LlamxbZsCLwg6vo1UACMFm3IALz+cgQk5O0ANAMUwaEA5ABi-NJQU2PMTLiNzWDIxHBgcMjkSSlpvf16WYJVBeN6GXp3RGTIo9f3yArKoQBERqlKvMgAd1MSmInzGrzu5Ws5ChpmeENYeiK4LKbUuLwIsI6KIIyJYqwA9ATkAAVIzAKTEXgQKRAqACXAwSggBAKJrILQQMD5ASoACeAEliAAKIYUGj0KAASjy2WQAB9kMyUjBQBBiKdCFAuZQoCAtjs4AA6c45I2qpScYWiyDUPIyKW9AB8RFtRpIJxIUoKAAYStFcETSeSpBTkNRTChOSBoMAEMhaIYttSpCBeJs6QyCQAqbP4bPIACiUCg7hCGAwcE5pHzyFrZLDogADtqK8B2WG05sEHAlEo4LQVEai4gjOGIHRoMheDA3j8lSB2-qwHymygprXsHXkAW7gAeADKHnEkjQ2SdtYhyGFPb7tEQAhgIHIwoAbr3KKEeSMiBJROQqEnKAfzgEs4D5WVCiKR0ehdd8lE-IgpAPH9jApcwoC0AB+GEQD5GVkAPOpLwhG9eyUe8EEfZ9r3gz9IIEH9QBSEQAIlaAQLAiCzygmCXWZAQ0wBEBUJDDDsNw-CGOIndkGYBV-GQI1lIYOTaymZAjEwMJgC0IIwF1Gkf0TTY0xjadZzcPwMF0-TDKkUCUEaagmx2YBBxQAFFDHCBR2nYxoCNfMCUZZlWSXDkuX5IU9xJJ1hRXNdyAECA+RnRJklSDAmMeQDJQIkkFOVCBVRjDVbmQbUDL1A1dgKRKICKc1gEtaBrUUCd7SwPjXQnd0NT6L1fX9ZggA)

**Comment by jeje42** [comment]
Hi, I also have the same problem as @Fleuv here with typescript 3.9.7. Can I ask what is the status of this issue and why it is closed ?

**Comment by AndyOGo** [comment]
Same Problem here with `Array.prototype.find()`

**Comment by SephReed** [comment]
I'm confused as to why this is closed as well.

I'm looking at code where I personally know *implicitly* **exactly** what the potential types are.  I know for a fact that this error is incorrect, and is a bug.

I could try to post the type here... it's a big gnarly recursive infer thing that I use for proxies.  And proxies have been historically poorly covered by typescript, so that's something I'm used to.

But this issue shouldn't be closed.  I'm looking at a case right now that's very, very clearly a fault in typescript that takes me almost no cognitive effort to confirm logically.

**Comment by SephReed** [comment]
*tl;dr:  to get around this issue ->  flatten your types to fit through whichever generic is the pigeon-hole, then use "type predicates" to make them whole again*

----------

**Understanding the Problem**

Alright, so after poking around for a bit, I've come to figure out what the heck is going on.  Basically:

```
type Foo = { foo: boolean };
type Bar = { bar: number };

// this works just fine
const mappable = [] as Array<Foo | Bar>;
mappable.map(it => it);

// this one can not be mapped
const cannotMap = [] as (Foo[] | Bar[]);
cannotMap.map(it => it); 
// ^^^^^ ERROR: Each member of the union type ... has signatures, but none of those signatures are compatible with each other

// but you can use **non-generic** functions on either... kinda
mappable.forEach(it => null);
cannotMap.forEach(it => null);
// ^^^ ERROR: Parameter 'it' implicitly has an 'any' type.

// Basically (Array<Foo | Bar>) and (Foo[] | Bar[]) are very different things as far as typescript is concerned
```
[Typescript Playground Demo Link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nKBeKBvKAzBAuKAjBAGwgEMA7KAXwG4AoUSKAIRICdk183cyBXAWzwR2NWrQD04qMAAWASwDOUAO5xWAayUArXguCY5ZCLQDGcMnqj8SYMCTzEOAbQC6UEkoCCrViRAAeeEQAH2Y2AD46a1t7YgA6aIAKOX0kcKgUgEo6CSlZRShzaBNyKDI4fSErG0gAE1NzSxKycuAAWRtnNw8oRKDXKFCWVlds03JWjrAEm2TU9KzqKFyoAD119agAUQAlHYB5HdwtkhMZKwhBYUKMaRloXjI5c2lwaDiPqBkehTkAczIJGAvFYEAUABp8Lx9OUjDc7nAFNBfgCgSCwe5QVAzPw7MA5A5oMoUudSGdCrJhGJJFD9CA4LxsaVdNAAFSs2EAWj+ECMrDkJnZmEeJnxjUKFAgJOEHziUHUhlqJFo0TshLiWFYJzOc2Q6T4hEIY2akxsGrU2pkurSZV4hrGNI22z2h1wAAU2CR+BBgNcAOQpP0ZXGEAUpQggL49Up+8ggIMMCBxalSFi-EqGyOJby+AJBQZhVjhTLuMi1Xr9NxDNijTHQABuwkjtTkGAwwl5+nyZD+Sh6GDY7iUiYUJn5YH0BTMZBMwiMtSAA)

------------

**Subjective Digression**

So, obviously this is kind of ridiculous.  As far as any developer is concerned, it would be good enough to be able to map `cannotMap` and get *slightly false* typing on the items as `(item: Foo | Bar) => any`.  In reality, the function should be `((item: Foo) => any) | ((item: Bar) => any)`..... whew!  really saving us from ourselves!

And, unfortunately, I've seen enough CLOSED typescript issues to know that this is never going to be changed.  It's pedantic and dumb, so they'll say they're being accurate.  In some other thread they'll shirk the whole ["preserve runtime behavior of all JavaScript code" accuracy](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals) thing, and say that [doing so would be pedantic](https://github.com/Microsoft/TypeScript/issues/2521#issuecomment-111208319).  So we're just going to have to work around this one

----------

**The Fix**

So this sucks, but the best way to fix this issue is to just flatten your types, then create type checkers to turn them back into what they're suppose to be.  Something like this:

```
type Foo = { foo: boolean };
type Bar = { bar: number };

type FooBar = Foo | Bar;

function isFoo(checkMe: FooBar): checkMe is Foo {
    return "foo" in checkMe;
}

function isBar(checkMe: FooBar): checkMe is Bar {
    return "bar" in checkMe;
}

const lame = [] as Array<FooBar>;
lame.map((it) => {
    if (isFoo(it)) {
        it.foo
    } else {
        it.bar
    }
})
```
[Typescript Playground Link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nKBeKBvKAzBAuKAjBAGwgEMA7KAXwG4AoUSKAIRICdk183cyBXAWzwR2NWvXDR4cFuxRSoAH2Zs6tDLzIBjYAEs4FHQGcpACk0ALCJoDWAWQi4pMgJS4LVu9COwEaWlACoVghgXlYKACIsOAioHQp3G3s6SjF1LV19OMMZM0skhx9pNlcoRM9s5XZUf0Dg0PCoCLw2WPiy-M8UsU19Q2AoQhJ+aBQAbQBdKBJDKABBVlYSEAAeJzYAPjohkYA6fhIwExMdYGdkDb9AuIwoE+MEE7Pzmuvr093o2oDKKAhCQzQV5vAIfFqsb5UWiUZxAA)

It's a super ugly solution in practice.  You have to flatten entire complex types into something that never realistically exists just so it'll fit through the pigeonhole that is `ts(2349)`.  But it works, so be it.

----------

**Subjective Suggestion**

I don't expect anyone with sway to be reading this, but on the off-chance such a thing should happen, I have a suggestion for making everyone happy; something like this:

```
// tsconfig.json. NOT REAL CODE.  Copy and Paste at your own disappointment.
{
  "compilerOptions": {
    "genericUnionStyle": "entireGenerics" | "genericArguments" = "entireGenerics",
  }
}
```

Pretty self explanatory, but for completenesses sake:
* the default remains the same as it is today.  nothing changes for the standard user.
* the new union style would solve the issue of `Array<Foo | Bar>` being usable, while `(Foo[] | Bar[])` is totally broken

```
// in action, to union the following...
type noLongerBrokenType = (Map<void, null> | Map<number, string);

// you'd end up with this
type sameAsAbove = Map<void | number, null | string>;
```

And even though this isn't *perfectly accurate*, it being accurate wouldn't really change the way I code at all.  Either way I'm going to be checking types, but at least this way *I wouldn't have felt the need to spend a couple hours figuring out what was going on then rewriting my code to be worse just to fix it...*. and then making this wall of text.

Hope this helps someone. <3

**Comment by luisanton-io** [comment]
> _tl;dr: to get around this issue -> flatten your types to fit through whichever generic is the pigeon-hole, then use "type predicates" to make them whole again_
> 
> **Understanding the Problem**
> 
> Alright, so after poking around for a bit, I've come to figure out what the heck is going on.
> 
> Basically `(Array<Foo | Bar>)` and `(Foo[] | Bar[])` are very different things as far as typescript is concerned:
> 
> ```ts
> type Foo = { foo: boolean };
> type Bar = { bar: number };
> 
> // this works just fine
> const mappable = [] as Array<Foo | Bar>;
> mappable.map(it => it);
> 
> // this one can not be mapped
> const cannotMap = [] as (Foo[] | Bar[]);
> cannotMap.map(it => it); 
> // ^^^^^ ERROR: Each member of the union type ... has signatures, but none of those signatures are compatible with each other
> 
> // but you can use **non-generic** functions on either... kinda
> mappable.forEach(it => null);
> cannotMap.forEach(it => null);
> // ^^^ ERROR: Parameter 'it' implicitly has an 'any' type.
> ```
> 
> [Typescript Playground Demo Link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nKBeKBvKAzBAuKAjBAGwgEMA7KAXwG4AoUSKAIRICdk183cyBXAWzwR2NWrQD04qMAAWASwDOUAO5xWAayUArXguCY5ZCLQDGcMnqj8SYMCTzEOAbQC6UEkoCCrViRAAeeEQAH2Y2AD46a1t7YgA6aIAKOX0kcKgUgEo6CSlZRShzaBNyKDI4fSErG0gAE1NzSxKycuAAWRtnNw8oRKDXKFCWVlds03JWjrAEm2TU9KzqKFyoAD119agAUQAlHYB5HdwtkhMZKwhBYUKMaRloXjI5c2lwaDiPqBkehTkAczIJGAvFYEAUABp8Lx9OUjDc7nAFNBfgCgSCwe5QVAzPw7MA5A5oMoUudSGdCrJhGJJFD9CA4LxsaVdNAAFSs2EAWj+ECMrDkJnZmEeJnxjUKFAgJOEHziUHUhlqJFo0TshLiWFYJzOc2Q6T4hEIY2akxsGrU2pkurSZV4hrGNI22z2h1wAAU2CR+BBgNcAOQpP0ZXGEAUpQggL49Up+8ggIMMCBxalSFi-EqGyOJby+AJBQZhVjhTLuMi1Xr9NxDNijTHQABuwkjtTkGAwwl5+nyZD+Sh6GDY7iUiYUJn5YH0BTMZBMwiMtSAA)
> 
> **Subjective Digression**
> 
> So, obviously this is kind of ridiculous. As far as any developer is concerned, it would be good enough to be able to map `cannotMap` and get _slightly false_ typing on the items as `(item: Foo | Bar) => any`. In reality, the function should be `((item: Foo) => any) | ((item: Bar) => any)`..... whew! really saving us from ourselves!
> 
> And -- unfortunately -- I've seen enough CLOSED typescript issues to know that this is never going to be changed. It's pedantic and dumb, so they'll say it's for the sake of not being inaccurate. Then on some other issue they'll shirk the whole ["preserve runtime behavior of all JavaScript code" accuracy](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals) thing, and say that [being an actual superset of javascript is too pedantic](https://github.com/Microsoft/TypeScript/issues/2521#issuecomment-111208319). So we're just going to have to work around this one.
> 
> **The Fix**
> 
> This sucks, BUT: the best way to fix this issue is to just flatten your types, then create type checkers to turn them back into what they're suppose to be. Something like the following:
> 
> ```ts
> type Foo = { foo: boolean };
> type Bar = { bar: number };
> 
> type FooBar = Foo | Bar;
> 
> function isFoo(checkMe: FooBar): checkMe is Foo {
>     return "foo" in checkMe;
> }
> 
> function isBar(checkMe: FooBar): checkMe is Bar {
>     return "bar" in checkMe;
> }
> 
> const lame = [] as Array<FooBar>;
> lame.map((it) => {
>     if (isFoo(it)) {
>         it.foo
>     } else {
>         it.bar
>     }
> })
> ```
> 
> [Typescript Playground Link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nKBeKBvKAzBAuKAjBAGwgEMA7KAXwG4AoUSKAIRICdk183cyBXAWzwR2NWvXDR4cFuxRSoAH2Zs6tDLzIBjYAEs4FHQGcpACk0ALCJoDWAWQi4pMgJS4LVu9COwEaWlACoVghgXlYKACIsOAioHQp3G3s6SjF1LV19OMMZM0skhx9pNlcoRM9s5XZUf0Dg0PCoCLw2WPiy-M8UsU19Q2AoQhJ+aBQAbQBdKBJDKABBVlYSEAAeJzYAPjohkYA6fhIwExMdYGdkDb9AuIwoE+MEE7Pzmuvr093o2oDKKAhCQzQV5vAIfFqsb5UWiUZxAA)
> 
> It's a super ugly solution in practice. You have to flatten entire complex types into something that never realistically exists just so it'll fit through the pigeonhole that is `ts(2349)`. But it works, so be it.
> 
> **Subjective Suggestion**
> 
> I don't expect anyone with sway to be reading this, but on the off-chance such a thing should happen, I have a suggestion for making everyone happy; something like this:
> 
> ```js
> // tsconfig.json. NOT REAL CODE.  Copy and Paste at your own disappointment.
> {
>   "compilerOptions": {
>     "genericUnionStyle": "entireGenerics" | "genericArguments" = "entireGenerics",
>   }
> }
> ```
> 
> Pretty self explanatory, but for completenesses sake:
> 
> * the default remains the same as it is today.  nothing changes for the standard user.
> * the new union style would solve the issue of `Array<Foo | Bar>` being usable, while `(Foo[] | Bar[])` is totally broken
> 
> ```ts
> // in action, to union the following...
> type noLongerBrokenType = (Map<void, null> | Map<number, string);
> 
> // you'd end up with this
> type sameAsAbove = Map<void | number, null | string>;
> ```
> 
> And even though this isn't _perfectly accurate_, it being accurate wouldn't really change the way anyone codes AFAICT. Either way you're going to have to check types. And given that the current way things are you can't even get to the point of checking types, it's not really helping anyone.
> 
> So why not give us the option to be able to move forwards at a tiny fraction of loss of accuracy? Surely that's better than everyone having to cannibalize their types just to put them back together on the other side of this error.
> 
> Hope this helps someone. <3

Thank you so much!

**Comment by bombillazo** [comment]
@luisanton-io Marvelous! That worked for me.

**Comment by aiham** [comment]
For my use case `(A | B)[]` and `(A[] | B[])` are equivalent so I was happy to workaround it with:

```typescript
const items = generateVariousItems();
const typedItems: ((typeof items)[number])[] = items;
const result = typedItems.find(item => item.commonField === 'foobar');
```

This way the types remain inferred from `generateVariousItems` instead of redefining them.

**Comment by howtomake** [comment]
you can use 
`_list: Array<type1 | type2> = []`

**Comment by scffs** [comment]
I was really surprised when I fix this error by replace `A[] | B[]` with `Array<A | B>`

**Comment by nicknish** [comment]
Hey folks, stumbled upon the same issue. I believe this is fixed in TypeScript 5.2:

https://devblogs.microsoft.com/typescript/announcing-typescript-5-2/#easier-method-usage-for-unions-of-arrays

**Comment by tejas-1206** [comment]
> I was really surprised when I fix this error by replace `A[] | B[]` with `Array<A | B>`

But how ? How is `Array<A | B>` different than `A[] | B[]` ? Does `Array<A> | Array<B>` work ?

**Comment by luisanton-io** [comment]
> > I was really surprised when I fix this error by replace `A[] | B[]` with `Array<A | B>`
> 
> But how ? How is `Array<A | B>` different than `A[] | B[]` ? Does `Array<A> | Array<B>` work ?

Yes, they are indeed different types.

```typescript
const arrayAorB: Array<A | B> = [a, b, a, b , b, ..., a]
const arrayA: Array<A> /* or: A[] */ = [a, a, a, a, ..., a]
const arrayB: Array<B> /* or: B[] */ = [b, b, b, b, ..., b]
const arrayAorArrayB: Array<A> | Array<B> /* or: A[] | B[] */ = [a, a, a, ..., a] | [b, b, b, ..., b]
```
